[BugFix] Unable to build from source, fix cmake error and use correct dependencies specified by third-party, clean unused variable.

matchyc opened the pull request:
## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes:
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #23111 

## Problem Summary(Required):
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->

TLDR; 
Users can't build Starrocks from the source.  And Starrocks is now using incorrect dependencies which does not follow the restriction.

### All the libraries mentioned below are now used in starrocks in a wrong way (not the restricted version), starrocks is searching on users local environment instead of using specified third-party, which may cause unknown behavior.

**Furthermore, even if we successfully pack the project for distribution, the several packages inside the libraries may come from the "packing machine", not the wanted versions of the library indicated in Starrocks' third-party restrictions.**

Mainly because the wrong build script and a little unused variable.

More **IMPORTATNT**, this PR enable Starrocks use correct and specified dependencies indicated by third-party file. 

Before this, many packages are used from system or python environment. 

In issue #23111 , I described the situation in detail, including how to locate the bug and a description of the solution.
Errors:
1. CMake Error: Could NOT find Snappy (missing: Snappy_LIB)
2. CMakeError: Count NOT find RapidJSONAlt
3. Could NOT find Lz4 (missing: LZ4_LIB LZ4_INCLUDE_DIR)
4. Not using zlib in third-party
5. Brotli is also using library in the environment 
6. Copy files missing
7. Declared but not used variable




If a user can build from the source, then these dependencies are installed manually by apt on Ubuntu or yum on CentOS, which is strongly not recommended.


## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

Commit:
fix build error while building third-party dependencies and clear unused variable to pass build. issue:#23111

Signed-off-by: matchyc <dawnlight.yc@protonmail.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/StarRocks/starrocks?pullRequest=23112) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/StarRocks/starrocks?pullRequest=23112) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/StarRocks/starrocks?pullRequest=23112) it.</sub>

kevincai commented:
those dependencies are not supposed to be manually installed by `apt` or `yum`, we have strict version control of the thirdparty dependencies.

please refer to doc: https://docs.starrocks.io/en-us/main/developers/build-starrocks/Build_in_docker, suggest following the instructions from the doc and try again.

kevincai commented on be/src/exprs/agg/stream/retract_maxmin.h:
```diff
@@ -236,10 +236,13 @@ class MaxMinAggregateFunctionRetractable final
         }
 
         // sync previous records from detail state table
+        // never used variable for now
```
let's delete the code if not used

kevincai commented on thirdparty/build-thirdparty.sh:
```diff
@@ -557,6 +557,20 @@ build_flatbuffers() {
+    cd $TP_SOURCE_DIR/$BROTLI_SOURCE
+    mkdir -p $BUILD_DIR
+    cd $BUILD_DIR
+    rm -rf CMakeCache.txt CMakeFiles/
```
thirdparty is supposed to be built from a brand new env. this clean up is not necessary.

kevincai commented on thirdparty/build-thirdparty.sh:
```diff
@@ -1110,8 +1132,9 @@ build_brpc
 build_flatbuffers
-# must build before arrow
 build_jemalloc
+build_brotli
```
brotli is built by arrow, we copied that after arrow is built, check the code line 621~624.

so we may not need to build it explicitly.

matchyc commented:
> those dependencies are not supposed to be manually installed by `apt` or `yum`, we have strict version control of the thirdparty dependencies.
> 
> please refer to doc: https://docs.starrocks.io/en-us/main/developers/build-starrocks/Build_in_docker, suggest following the instructions from the doc and try again.

That is exactly what I fixed in this PR. This PR is actually fix the version control. 
Before: can't build and library will be found on users local environments.
Fixed: Use starrocks speficied third-party dependencies. 
Please see detail description in issue #23111 . I described how the bug occurs and the correspond solution.

kevincai commented on thirdparty/build-thirdparty.sh:
```diff
@@ -568,6 +582,7 @@ build_arrow() {
 
     check_if_source_exist $ARROW_SOURCE
     cd $TP_SOURCE_DIR/$ARROW_SOURCE/cpp
+    rm -rf CMakeCache.txt CMakeFiles/
```
ditto

imay commented:
@matchyc could you sign the CAL?

matchyc commented:
> @matchyc could you sign the CAL?

Sure, signed it.

kevincai commented on thirdparty/build-thirdparty.sh:
```diff
@@ -566,6 +566,19 @@ build_flatbuffers() {
+    cd $TP_SOURCE_DIR/$BROTLI_SOURCE
+    mkdir -p $BUILD_DIR
+    cd $BUILD_DIR
+    ${CMAKE_CMD} .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=$TP_INSTALL_DIR
+    ${CMAKE_CMD} --build . --config Release --target install
```
```suggestion
    ${CMAKE_CMD} .. -G "${CMAKE_GENERATOR}" -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=$TP_INSTALL_DIR
    ${BUILD_SYSTEM} -j$PARALLEL
    ${BUILD_SYSTEM} install
```

kevincai replied:
if we can keep consistent with other dependency build command. support both make and ninja and takes `$PARALLEL` in consideration.

kevincai commented on thirdparty/build-thirdparty.sh:
```diff
@@ -566,6 +566,19 @@ build_flatbuffers() {
+    ${CMAKE_CMD} .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=$TP_INSTALL_DIR
+    ${CMAKE_CMD} --build . --config Release --target install
+
+    cp -rf $TP_INSTALL_DIR/lib/libbrotlienc-static.a $TP_INSTALL_DIR/lib64/libbrotlienc.a
+    cp -rf $TP_INSTALL_DIR/lib/libbrotlidec-static.a $TP_INSTALL_DIR/lib64/libbrotlidec.a
+    cp -rf $TP_INSTALL_DIR/lib/libbrotlicommon-static.a $TP_INSTALL_DIR/lib64/libbrotlicommon.a
```
I would rather use `mv` than `cp`, don't want multiple copies in the target install location.

kevincai replied:
also need to remove line 629~631, since we do it separately, so no need to ask arrow to do it again.

kevincai commented on be/src/exprs/agg/stream/retract_maxmin.h:
```diff
@@ -236,10 +236,7 @@ class MaxMinAggregateFunctionRetractable final
-        if constexpr (!lt_is_string<LT>) {
-            column = down_cast<InputColumnType*>(columns[0].get());
-        }
+
```
remove this empty line.

kevincai reviewed (commented):
overall LGTM. let's check the CI job, if everything goes as exepcted. (https://github.com/StarRocks/starrocks/actions/runs/4936858122/jobs/8824929321?pr=23112)

kevincai commented on thirdparty/build-thirdparty.sh:
```diff
@@ -566,6 +566,20 @@ build_flatbuffers() {
+    ${BUILD_SYSTEM} -j$PARALLEL
+    ${BUILD_SYSTEM} install
+
+    mv -f $TP_INSTALL_DIR/lib/libbrotlienc-static.a $TP_INSTALL_DIR/lib64/libbrotlienc.a
```
![image](https://github.com/StarRocks/starrocks/assets/771299/49d53338-9aa7-497b-903e-d706c245ea7c)

kevincai replied:
1. need disable shared library build, only build static library
2. installed to `lib64` location by default on centos7, but possibly to `lib` location on ubuntu, need specify cmake_lib_dir to make the target library location predictable.

matchyc commented:
It seems that the building test runs well. Note that in the latest commit. I also fixed that arrow is now using the correct boost library in third-party instead of using the environment boost library.

kevincai commented:
![image](https://github.com/StarRocks/starrocks/assets/771299/1da6bdbb-2dcc-47e8-ae5b-f842a7e30264)

looks like still can't find the rapidjson?

matchyc commented:
> ![image](https://user-images.githubusercontent.com/771299/237566539-1da6bdbb-2dcc-47e8-ae5b-f842a7e30264.png)
> 
> looks like still can't find the rapidjson?

After checking this warning carefully, I found it meaningless. I use `RapidJSON_ROOT` to indicate the path and ensure starrocks will use the specific third-party library, but before that:

In arrow cmake_module files `FindRapidJSONAlt.cmake`
```cmake
set(find_package_args)
if(RapidJSONAlt_FIND_VERSION)
  list(APPEND find_package_args ${RapidJSONAlt_FIND_VERSION})
endif()
if(RapidJSONAlt_FIND_QUIETLY)
  list(APPEND find_package_args QUIET)
endif()
find_package(RapidJSON ${find_package_args}) # note here!!!
if(RapidJSON_FOUND)
  set(RapidJSONAlt_FOUND TRUE)
  set(RAPIDJSON_INCLUDE_DIR ${RAPIDJSON_INCLUDE_DIRS})
  return()
endif()

if(RapidJSON_ROOT)
  find_path(RAPIDJSON_INCLUDE_DIR
            NAMES rapidjson/rapidjson.h
            PATHS ${RapidJSON_ROOT}
            NO_DEFAULT_PATH
            PATH_SUFFIXES "include")
else()
  find_path(RAPIDJSON_INCLUDE_DIR
            NAMES rapidjson/rapidjson.h
            PATH_SUFFIXES "include")
endif()
```
This means, arrow triggers find_package to find RapidJSON, which will do nothing. Eventually, it will find RapidJSON in the `RapidJSON_ROOT` directory. 
And we can't modify this because it is inside the arrow project. Except for the warning, everything goes well as expected.

kevincai commented on thirdparty/build-thirdparty.sh:
```diff
@@ -566,6 +566,19 @@ build_flatbuffers() {
+    ${CMAKE_CMD} .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=$TP_INSTALL_DIR -DBROTLI_EMSCRIPTEN=ON -DCMAKE_INSTALL_LIBDIR=lib64
+    ${BUILD_SYSTEM} -j$PARALLEL
+
+    cp -f libbrotlienc-static.a $TP_INSTALL_DIR/lib64/libbrotlienc.a
```
one last question, why use cp instead of `${BUILD_SYSTEM} install`

kevincai commented:
have manual tried with a complete log. looks like arrow is still trying to rebuild brotli_ep itself.

![image](https://github.com/StarRocks/starrocks/assets/771299/34539d88-c0bc-4db3-8725-154aea257b02)

If this is the case, looks like we are doing for nothing building the brotli outside arrow project.

matchyc commented:
> have manual tried with a complete log. looks like arrow is still trying to rebuild brotli_ep itself.
> 
> ![image](https://user-images.githubusercontent.com/771299/237860428-34539d88-c0bc-4db3-8725-154aea257b02.png)
> 
> If this is the case, looks like we are doing for nothing building the brotli outside arrow project.

That's weird because it is not suited for the logic of code. Checking the CMakelists and cmake_modules provided by arrow, we know it won't be built unless specific args were set. Like options for `xsimd` is set to "BUNDLE" by hard code. That is why `zstd` and `xsimd` will be built by arrow. 
```cmake
# this file is from arrow
if(...)
  set(xsimd_SOURCE "BUNDLED") # note this !!!!
 # ...
endif()
```
But there is no flag for `brotli`. And in my local Ubuntu machine with cmake version 3.26 or 3.16, `brotli` will always be found.
```
-- Found Brotli: <my_path>/starrocks/thirdparty/installed/lib64/libbrotlicommon.a  
-- Providing CMake module for Brotli
```

Could you provide the commit_id you use to reproduce the procedure? The latest one in this pr is `69e38e1206496c1bf390bd5b6bde49bcfcfa8512`

kevincai commented:
> > have manual tried with a complete log. looks like arrow is still trying to rebuild brotli_ep itself.
> > ![image](https://user-images.githubusercontent.com/771299/237860428-34539d88-c0bc-4db3-8725-154aea257b02.png)
> > If this is the case, looks like we are doing for nothing building the brotli outside arrow project.
> 
> That's weird because it is not suited for the logic of code. Checking the CMakelists and cmake_modules provided by arrow, we know it won't be built unless specific args were set. Like options for `xsimd` is set to "BUNDLE" by hard code. That is why `zstd` and `xsimd` will be built by arrow.
> 
> ```cmake
> # this file is from arrow
> if(...)
>   set(xsimd_SOURCE "BUNDLED") # note this !!!!
>  # ...
> endif()
> ```
> 
> But there is no flag for `brotli`. And in my local Ubuntu machine with cmake version 3.26 or 3.16, `brotli` will always be found.
> 
> ```
> -- Found Brotli: <my_path>/starrocks/thirdparty/installed/lib64/libbrotlicommon.a  
> -- Providing CMake module for Brotli
> ```
> 
> Could you provide the commit_id you use to reproduce the procedure? And I can check deep in, or the full log for the arrow building part. There are library finding in formation above the messages in this picture. The latest one in this pr is `69e38e1206496c1bf390bd5b6bde49bcfcfa8512`

I uses exactly this PR's commit `gh checkout pr 23112`

Reproduce it docker build command as follows: 
```
DOCKER_BUILDKIT=1 docker build --build-arg predownload_thirdparty=true --build-arg prebuild_maven=false -f docker/dockerfiles/dev-env/dev-env.Dockerfile --build-arg distro=centos7 -t nobody/dev-env-centos7:main-pr23112-2 .
```

to produce a centos7 env build.

kevincai commented:
```
#12 631.4 -- Looking for pthread.h
#12 631.5 -- Looking for pthread.h - found
#12 631.5 -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
#12 631.6 -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
#12 631.6 -- Found Threads: TRUE  
#12 631.6 -- Found Thrift: /root/starrocks/thirdparty/installed/lib/libthrift.a (found suitable version "0.13.0", minimum required is "0.11.0")  
#12 631.6 -- Looking for __SIZEOF_INT128__
#12 631.7 -- Looking for __SIZEOF_INT128__ - found
#12 631.7 -- Found Snappy: /root/starrocks/thirdparty/installed/lib64/libsnappy.a  
#12 631.7 -- Providing CMake module for Snappy
#12 631.7 -- Could NOT find Brotli (missing: BROTLI_INCLUDE_DIR) 
#12 631.7 -- Building brotli from source
#12 631.7 -- Building without OpenSSL support. Minimum OpenSSL version 1.0.2 required.
#12 631.7 -- Building (vendored) jemalloc from source
```

I guess most likely because the `brotli` include file is not installed?

matchyc commented:
> > > have manual tried with a complete log. looks like arrow is still trying to rebuild brotli_ep itself.
> > > ![image](https://user-images.githubusercontent.com/771299/237860428-34539d88-c0bc-4db3-8725-154aea257b02.png)
> > > If this is the case, looks like we are doing for nothing building the brotli outside arrow project.
> > 
> > 
> > That's weird because it is not suited for the logic of code. Checking the CMakelists and cmake_modules provided by arrow, we know it won't be built unless specific args were set. Like options for `xsimd` is set to "BUNDLE" by hard code. That is why `zstd` and `xsimd` will be built by arrow.
> > ```cmake
> > # this file is from arrow
> > if(...)
> >   set(xsimd_SOURCE "BUNDLED") # note this !!!!
> >  # ...
> > endif()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But there is no flag for `brotli`. And in my local Ubuntu machine with cmake version 3.26 or 3.16, `brotli` will always be found.
> > ```
> > -- Found Brotli: <my_path>/starrocks/thirdparty/installed/lib64/libbrotlicommon.a  
> > -- Providing CMake module for Brotli
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Could you provide the commit_id you use to reproduce the procedure? And I can check deep in, or the full log for the arrow building part. There are library finding in formation above the messages in this picture. The latest one in this pr is `69e38e1206496c1bf390bd5b6bde49bcfcfa8512`
> 
> I uses exactly this PR's commit `gh checkout pr 23112`
> 
> Reproduce it docker build command as follows:
> 
> ```
> DOCKER_BUILDKIT=1 docker build --build-arg predownload_thirdparty=true --build-arg prebuild_maven=false -f docker/dockerfiles/dev-env/dev-env.Dockerfile --build-arg distro=centos7 -t nobody/dev-env-centos7:main-pr23112-2 .
> ```
> 
> to produce a centos7 env build.

Tried several times but can't reproduce. Docker (docker 23.0.2) didn't give the expected output (some are omitted even --progress=plain), would you like to help to try adding a definition like ` -DBROTLI_INCLUDE_DIR=$TP_INSTALL_DIR/include/brotli` to see if it goes well?  Because that's hard for me to debug this if I can't reproduce it.

matchyc commented:
If it does, then that may be different behaviors between os? But may only give benefit because without building brotli it can't be built from source indeed in `ubuntu 20.04 5.15.0-56-generic`

kevincai commented:
> If it does, then that may be different behaviors between os? But may only give benefit because without building brotli it can't be built from source indeed in `ubuntu 20.04 5.15.0-56-generic`

It is not specific on centos7, I tried both centos7 and ubuntu22.04.

kevincai commented:
> > > > have manual tried with a complete log. looks like arrow is still trying to rebuild brotli_ep itself.
> > > > ![image](https://user-images.githubusercontent.com/771299/237860428-34539d88-c0bc-4db3-8725-154aea257b02.png)
> > > > If this is the case, looks like we are doing for nothing building the brotli outside arrow project.
> > > 
> > > 
> > > That's weird because it is not suited for the logic of code. Checking the CMakelists and cmake_modules provided by arrow, we know it won't be built unless specific args were set. Like options for `xsimd` is set to "BUNDLE" by hard code. That is why `zstd` and `xsimd` will be built by arrow.
> > > ```cmake
> > > # this file is from arrow
> > > if(...)
> > >   set(xsimd_SOURCE "BUNDLED") # note this !!!!
> > >  # ...
> > > endif()
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > But there is no flag for `brotli`. And in my local Ubuntu machine with cmake version 3.26 or 3.16, `brotli` will always be found.
> > > ```
> > > -- Found Brotli: <my_path>/starrocks/thirdparty/installed/lib64/libbrotlicommon.a  
> > > -- Providing CMake module for Brotli
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Could you provide the commit_id you use to reproduce the procedure? And I can check deep in, or the full log for the arrow building part. There are library finding in formation above the messages in this picture. The latest one in this pr is `69e38e1206496c1bf390bd5b6bde49bcfcfa8512`
> > 
> > 
> > I uses exactly this PR's commit `gh checkout pr 23112`
> > Reproduce it docker build command as follows:
> > ```
> > DOCKER_BUILDKIT=1 docker build --build-arg predownload_thirdparty=true --build-arg prebuild_maven=false -f docker/dockerfiles/dev-env/dev-env.Dockerfile --build-arg distro=centos7 -t nobody/dev-env-centos7:main-pr23112-2 .
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to produce a centos7 env build.
> 
> Tried several times but can't reproduce. Docker (docker 23.0.2) didn't give the expected output (some are omitted even --progress=plain), would you like to help to try adding a definition like ` -DBROTLI_INCLUDE_DIR=$TP_INSTALL_DIR/include/brotli` during arrow building to see if it goes well? Because that's hard for me to debug this if I can't reproduce it.

you can comment out some unrelated build packages, e.g. aws sdk cpp, to save the output.

matchyc commented:
2 approving reviews are needed

kevincai commented:
> 2 approving reviews are needed

@chaoyli @imay please take a look.
